src/app/services/keycloak.service.ts

```diff
@@ -74,134 +63,168 @@ export class KeycloakService {
-          console.log('onAuthRefreshError');
+        self.keycloakAuth.onAuthRefreshError = () => {
           self.keycloakAuth.login({ idpHint: 'idir' });
         };
 
-        self.keycloakAuth.onAuthLogout = function () {
-          // console.log('onAuthLogout');
-        };
-
         // Try to get refresh tokens in the background
-        self.keycloakAuth.onTokenExpired = function () {
+        self.keycloakAuth.onTokenExpired = () => {
           self.keycloakAuth.updateToken(180)
-            .success(function (refreshed) {
-              console.log('KC refreshed token?:', refreshed);
-            })
-            .error((err) => {
-              console.log('onTokenExpired:KC refresh error:', err);
+            .success()
+            .error(() => {
               self.keycloakAuth.login({ idpHint: 'idir' });
             });
         };
 
-        // Initialize.
-        self.keycloakAuth.init({})
+        self.keycloakAuth.init({
+          // Specify the pkceMethod to be compatible with Common Hosted SSO.
+          pkceMethod: 'S256',
+          onLoad: 'check-sso',
+          // Without this option, keycloak assumes we're logged out and dumps all tokens.
+          checkLoginIframe: false
+        })
           .success((auth) => {
-            // console.log('KC Refresh Success?:', self.keycloakAuth.authServerUrl);
-            console.log('KC Success:', auth);
-
-            // After successful login, see if there's a user model for project permissions
-            this.checkUser(self.keycloakAuth.tokenParsed);
-
             if (!auth) {
-              if (this.loggedOut === 'true') {
+              if ('true' === this.loggedOut) {
                 // Don't do anything, they wanted to remain logged out.
                 resolve();
               } else {
                 self.keycloakAuth.login({ idpHint: 'idir' });
               }
             } else {
+              const userToken = self.keycloakAuth.tokenParsed;
+              window.localStorage.setItem('currentUser', JSON.stringify({ username: userToken.displayName, token: this.keycloakAuth.token }));
+              // After successful login, see if there's a user model for project permissions
+              this.checkUser(userToken)
+                .subscribe(
+                  userArray => {
+                    if (0 === userArray.length) {
+                      // If no user exists with this email, add them to the DB.
+                      this.addNewUser(userToken)
+                        .subscribe(() => {
+                          resolve();
+                        }, err => {
+                          console.error('Could not add new user', err);
+                          reject();
+                        });
+                    } else if (1 === userArray.length) {
+                      const user = userArray[0];
+                      if (user.idirUserGuid && user.idirUserGuid === userToken.idir_user_guid) {
```

Comment:
Good point. The idirUserGuid is set in the app code so that should be pretty certain. As for the `userToken.idir_user_guid`, it's set by the Shared SSO Service. So maybe putting in a check for that would be prudent. Still, you'd expect them to announce if they ever planned to change that since I imagine many govt apps must rely on it.